modules/mastodon/index.js

```diff
@@ -0,0 +1,320 @@
+  if (record.type == 'Note') {
+    fetchResource(record.attributedTo, (error, userRecord) => {
+      var str = '';
+      if (error) {
+        str += '(error getting user): ';
+      } else {
+        str += `${userRecord.name} (@${userRecord.preferredUsername}): `;
+      }
+      if (record.sensitive) {
+        str += sprintf('[content warning] %s', record.summary);
+      } else {
+        str += htmlToText(record.content);
+      }
+      fn(str);
+    });
+  } else if (record.type == 'Person') {
+    fn(`${record.name} (@${record.preferredUsername}): ` + htmlToText(record.summary));
+  }
+}
+
+module.exports.commands = ['masto'];
+
+module.exports.init = function(b) {
+  bot = b;
+  log = bot.log;
+
+  bot.readDataFile('config.json', function(err, data) {
+    if (err) {
+      log.info('initializing mastodon module data');
+      moduleConfig = defaultConfig();
+    } else {
+      try {
+        moduleConfig = JSON.parse(data);
+      } catch(ex) {
+        moduleConfig = defaultConfig();
+      }
+    }
+    writeConfig();
+  });
+
+  var userAgent = moduleConfig && moduleConfig.userAgent || 'EuIrcBot';
+  try {
+    var ver = require('../../package.json').version;
+    userAgent += '/' + ver;
+  } catch (ex) {
+    log.error('error getting package version: ' + ex);
+  }
+
+  var userAgentUrl = (moduleConfig && moduleConfig.userAgentUrl
+    || 'https://github.com/euank/EuIrcBot/blob/{commitish}/modules/mastodon/info.md');
+  if (userAgentUrl.includes('{commitish}')) {
+    var commitish = 'master';
+    try {
+      commitish = execSync('git rev-parse --short HEAD', {'encoding': 'utf8'}).trim();
+    } catch (ex) {
+      log.error('error determining deployed commit: ' + ex);
+    }
+    userAgentUrl = userAgentUrl.replace(/{commitish}/g, commitish);
+  }
+  userAgent += ' (+' + userAgentUrl + ')';
+
+  request = require('request').defaults({
+    headers: {'User-Agent': userAgent}
+  });
+};
+
+function handleUrl(url, reply) {
+  fetchResource(url, (error, record) => {
+    if (error) {
+      error.mastodonUrl = url;
+      log.info(error);
+      reply('error: ' + error.message);
+    } else {
+      formatRecord(record, (s) => reply.custom({ replaceNewlines: true }, s));
+    }
+  });
+
+  /*
+   * Instance autodiscovery: hit /api/v1/timelines/public and look for any new
+   * instance domains, then run those through forceCheckHostname to see if they
+   * ought to be added to our list.
+   *
+   * This will cause Pleroma instances to be added to the whitelist, even
+   * though they don't support ActivityPub objects, because they support the
+   * Mastodon client API.
+   */
+
+  robotRequest(
+    { url: urlMod.resolve(url, '/api/v1/timelines/public') },
+    (error, response, body) => {
+      if (error || response.statusCode != 200)
+        return;
+
+      try {
+        body = JSON.parse(body);
+      } catch (e) {
+        return;
+      }
+
+      arrayUniq(body.map((item) => urlMod.parse(item.url))).forEach((url) => {
```

Comment:
This doesn't work as written since two "Url" objects (and really two objects in js ever) aren't equal. [Example](https://runkit.com/euank/5a14dce8e5e2d900128e1952).

Since we only care about `host` anyways, we should just grab that out in the map, and then uniq should work, but we don't have to pull in a dependency on it with es6, we can use a set!

Something like [this](https://runkit.com/euank/5a14ddc9f74137001253c1e9) should work.

Reply:
Oh, right. I had the `arrayUniq` before I realized I needed the whole URL object to pass to `forceCheckHostname`...

honestly, I should just figure out how to make `forceCheckHostname` take the right args. It's done nothing but confuse me.